sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvcompLZ4CompressionCodec.scala

```diff
@@ -0,0 +1,169 @@
+
+  /**
+   * Compress a data buffer.
+   * @param input buffer containing data to compress
+   * @param stream CUDA stream to use
+   * @return the size of the compressed data in bytes and the (probably oversized) output buffer
+   */
+  def compress(input: DeviceMemoryBuffer, stream: Cuda.Stream): (Long, DeviceMemoryBuffer) = {
+    val tempSize = LZ4Compressor.getTempSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE)
+    withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+      var compressedSize: Long = 0L
+      val outputSize = LZ4Compressor.getOutputSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+        tempBuffer)
+      closeOnExcept(DeviceMemoryBuffer.allocate(outputSize)) { outputBuffer =>
+        compressedSize = LZ4Compressor.compress(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+          tempBuffer, outputBuffer, stream)
+        require(compressedSize <= outputBuffer.getLength, "compressed buffer overrun")
+        (compressedSize, outputBuffer)
+      }
+    }
+  }
+
+  /**
+   * Decompress data asynchronously that was compressed with nvcomp's LZ4-GPU codec
+   * @param outputBuffer where the uncompressed data will be written
+   * @param inputBuffer buffer of compressed data to decompress
+   * @param stream CUDA stream to use
+   */
+  def decompressAsync(
+      outputBuffer: DeviceMemoryBuffer,
+      inputBuffer: DeviceMemoryBuffer,
+      stream: Cuda.Stream): Unit = {
+    withResource(Decompressor.getMetadata(inputBuffer, stream)) { metadata =>
+      val outputSize = Decompressor.getOutputSize(metadata)
+      if (outputSize != outputBuffer.getLength) {
+        throw new IllegalStateException(
+          s"metadata uncompressed size is $outputSize, buffer size is ${outputBuffer.getLength}")
+      }
+      val tempSize = Decompressor.getTempSize(metadata)
+      withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+        Decompressor.decompressAsync(inputBuffer, tempBuffer, metadata, outputBuffer, stream)
+      }
+    }
+  }
+}
+
+class BatchedNvcompLZ4Compressor(maxBatchMemorySize: Long, stream: Cuda.Stream)
+    extends BatchedTableCompressor(maxBatchMemorySize, stream) {
+  override protected def compress(
+      tables: Array[ContiguousTable],
+      stream: Cuda.Stream): Array[CompressedTable] = {
+    val inputBuffers: Array[BaseDeviceMemoryBuffer] = tables.map(_.getBuffer)
+    val compressionResult = BatchedLZ4Compressor.compress(inputBuffers,
+      NvcompLZ4CompressionCodec.LZ4_CHUNK_SIZE, stream)
+    val compressedTables = try {
+      val buffers = compressionResult.getCompressedBuffers
+      val compressedSizes = compressionResult.getCompressedSizes
+      buffers.zipWithIndex.map { case (buffer, i) =>
+        val contigTable = tables(i)
+        val compressedSize = compressedSizes(i)
+        require(compressedSize <= buffer.getLength, "compressed buffer overrun")
+        val meta = MetaUtils.buildTableMeta(
+          tableId = 0,
```

Comment:
when `tableId=0`, something else is mutating this later?

Reply:
Yes, that mutation occurs in the RAPIDS shuffle at https://github.com/NVIDIA/spark-rapids/blob/dac7a91215e501abae61d49fb6dba6421e54563a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManager.scala#L121-L122